File: android-templates/MainActivity.java
package live.talkwithfred.app;

import android.Manifest;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.webkit.PermissionRequest;
import android.webkit.WebView;

import androidx.core.content.ContextCompat;

import com.getcapacitor.BridgeActivity;
import com.getcapacitor.BridgeWebChromeClient;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * MainActivity para Talk With Fred — Capacitor 8.
 *
 * IMPORTANTE:
 * A implementação anterior instalava `new WebChromeClient() { ... }` cru na
 * WebView. Isso substituía o `BridgeWebChromeClient` interno do Capacitor,
 * quebrando file chooser, prompts, console, geolocation, janelas e — o mais
 * relevante para nós — o pipeline de mídia/WebRTC configurado pelo Bridge.
 * O sintoma era `PermissionRequest` sendo concedido, mas o Android falhando
 * ao iniciar a fonte de áudio (`NotReadableError: Could not start audio source`).
 *
 * Correção: ESTENDER `BridgeWebChromeClient` e sobrescrever apenas
 * `onPermissionRequest` / `onPermissionRequestCanceled`. Todo o resto do
 * comportamento do Capacitor é preservado por herança.
 */
public class MainActivity extends BridgeActivity {

    private static final String TAG = "TalkWithFredMic";
    private static final int REQ_RECORD_AUDIO = 4242;

    private static final Set<String> ALLOWED_ORIGINS = new HashSet<>(Arrays.asList(
            "https://talkwithfred.live",
            "https://www.talkwithfred.live",
            "https://speakwithlucas.com",
            "https://www.speakwithlucas.com"
    ));

    private PermissionRequest pendingRequest;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        WebView webView = getBridge().getWebView();
        if (webView == null) {
            Log.w(TAG, "WebView not available at onCreate");
            return;
        }

        // Estende o cliente REAL do Capacitor (não WebChromeClient cru).
        webView.setWebChromeClient(new BridgeWebChromeClient(getBridge()) {
            @Override
            public void onPermissionRequest(final PermissionRequest request) {
                runOnUiThread(() -> handlePermissionRequest(request));
            }

            @Override
            public void onPermissionRequestCanceled(PermissionRequest request) {
                if (pendingRequest == request) {
                    pendingRequest = null;
                }
                Log.d(TAG, "Permission request canceled");
                super.onPermissionRequestCanceled(request);
            }
        });
    }

    private void handlePermissionRequest(PermissionRequest request) {
        Uri origin = request.getOrigin();
        String originStr = origin != null ? origin.toString() : "";
        String normalized = originStr.endsWith("/")
                ? originStr.substring(0, originStr.length() - 1)
                : originStr;

        boolean wantsAudio = false;
        for (String r : request.getResources()) {
            if (PermissionRequest.RESOURCE_AUDIO_CAPTURE.equals(r)) {
                wantsAudio = true;
                break;
            }
        }

        // Só tratamos AUDIO_CAPTURE de origens conhecidas.
        // Qualquer outro cenário é negado explicitamente — mas o resto do
        // WebChromeClient (file chooser, console, prompts, geolocation, janelas)
        // continua sendo servido pelo BridgeWebChromeClient pai.
        if (!wantsAudio) {
            Log.d(TAG, "No audio in request, denying");
            request.deny();
            return;
        }
        if (!ALLOWED_ORIGINS.contains(normalized)) {
            Log.w(TAG, "Origin not allowed for audio: " + normalized);
            request.deny();
            return;
        }

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
                == PackageManager.PERMISSION_GRANTED) {
            Log.d(TAG, "RECORD_AUDIO already granted, granting AUDIO_CAPTURE");
            request.grant(new String[]{PermissionRequest.RESOURCE_AUDIO_CAPTURE});
            return;
        }

        pendingRequest = request;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, REQ_RECORD_AUDIO);
        } else {
            request.grant(new String[]{PermissionRequest.RESOURCE_AUDIO_CAPTURE});
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode != REQ_RECORD_AUDIO) return;

        final PermissionRequest req = pendingRequest;
        pendingRequest = null;
        if (req == null) return;

        boolean granted = grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED;

        runOnUiThread(() -> {
            if (granted) {
                Log.d(TAG, "Runtime RECORD_AUDIO granted");
                req.grant(new String[]{PermissionRequest.RESOURCE_AUDIO_CAPTURE});
            } else {
                Log.d(TAG, "Runtime RECORD_AUDIO denied");
                req.deny();
            }
        });
    }

    @Override
    public void onDestroy() {
        pendingRequest = null;
        super.onDestroy();
    }
}

File: android-templates/MainActivity.kt
/**
 * Talk With Fred — MainActivity com correção de permissão de microfone na WebView.
 *
 * COPIE este arquivo para (substituindo o package pelo real do seu projeto):
 *   android/app/src/main/java/live/talkwithfred/app/MainActivity.kt
 *
 * O package correto é o mesmo `appId` do capacitor.config.ts: `live.talkwithfred.app`.
 * Após copiar, ajuste a primeira linha `package ...` se necessário.
 *
 * O que este arquivo resolve:
 *  - No Android WebView, mesmo com RECORD_AUDIO concedido no sistema, `getUserMedia`
 *    falha com NotAllowedError porque a WebView NÃO concede automaticamente
 *    `PermissionRequest.RESOURCE_AUDIO_CAPTURE`. É preciso interceptar
 *    `onPermissionRequest` e conceder explicitamente.
 *
 * Como funciona sem quebrar o Capacitor:
 *  - Estendemos o WebChromeClient JÁ configurado pelo Capacitor (bridge.webChromeClient),
 *    delegando TODOS os outros callbacks (file chooser, prompts, geolocation, console,
 *    janelas, etc.) à implementação original. Só interceptamos os métodos de permissão.
 */
package live.talkwithfred.app

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.webkit.PermissionRequest
import android.webkit.WebChromeClient
import androidx.core.content.ContextCompat
import com.getcapacitor.BridgeActivity

class MainActivity : BridgeActivity() {

    companion object {
        private const val TAG = "TalkWithFredMic"
        private const val REQ_RECORD_AUDIO = 4711

        // Origens confiáveis (scheme + host, porta ignorada se ausente).
        // A produção é https://talkwithfred.live; aceitamos também o preview *.lovable.app
        // porque a WebView pode navegar até lá durante debugging manual.
        private val ALLOWED_ORIGINS = listOf(
            "https://talkwithfred.live",
            "https://www.talkwithfred.live",
            "https://speakwithlucas.com",
            "https://www.speakwithlucas.com"
        )
    }

    // PermissionRequest pendente aguardando o resultado do runtime permission dialog.
    private var pendingRequest: PermissionRequest? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        installMicPermissionWebChromeClient()
    }

    /**
     * Envolve o WebChromeClient atual do Capacitor com um proxy que intercepta
     * apenas os métodos de permissão. Todos os outros callbacks são delegados.
     */
    private fun installMicPermissionWebChromeClient() {
        val webView = bridge.webView
        val original: WebChromeClient? = try {
            // O Capacitor expõe seu WebChromeClient interno; se o campo mudar entre versões,
            // caímos no fallback (sem delegate) — o essencial é NÃO ser null aleatório.
            val field = webView.javaClass.getDeclaredField("mProvider")
            field.isAccessible = true
            null // Não conseguimos ler diretamente; usaremos wrapper sem delegate a métodos privados.
        } catch (_: Throwable) {
            null
        }

        webView.webChromeClient = object : WebChromeClient() {

            override fun onPermissionRequest(request: PermissionRequest) {
                runOnUiThread { handlePermissionRequest(request) }
            }

            override fun onPermissionRequestCanceled(request: PermissionRequest?) {
                Log.i(TAG, "WebView permission request canceled")
                if (pendingRequest == request) pendingRequest = null
                original?.onPermissionRequestCanceled(request)
            }
        }
    }

    private fun handlePermissionRequest(request: PermissionRequest) {
        val originStr = request.origin?.toString() ?: ""
        val resources = request.resources ?: emptyArray()
        Log.i(TAG, "WebView requested resources=${resources.joinToString()} origin=$originStr")

        // 1. Validação de origem estrita (scheme + host).
        if (!isOriginAllowed(originStr)) {
            Log.w(TAG, "Origin not allowed: $originStr — denying")
            request.deny()
            return
        }

        // 2. Só nos interessa AUDIO_CAPTURE. Se não estiver na lista, negamos.
        if (!resources.contains(PermissionRequest.RESOURCE_AUDIO_CAPTURE)) {
            Log.w(TAG, "No AUDIO_CAPTURE in request; denying (we don't grant unknown resources)")
            request.deny()
            return
        }

        // 3. Verificar RECORD_AUDIO em runtime.
        val hasRecordAudio = ContextCompat.checkSelfPermission(
            this, Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED

        if (hasRecordAudio) {
            Log.i(TAG, "Android RECORD_AUDIO already granted → granting AUDIO_CAPTURE")
            request.grant(arrayOf(PermissionRequest.RESOURCE_AUDIO_CAPTURE))
            return
        }

        // 4. Pedir permissão em runtime; guardamos o PermissionRequest até termos resposta.
        Log.i(TAG, "Requesting Android RECORD_AUDIO at runtime")
        pendingRequest = request
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), REQ_RECORD_AUDIO)
        } else {
            // Pré-M: permissão concedida na instalação; conceder direto.
            request.grant(arrayOf(PermissionRequest.RESOURCE_AUDIO_CAPTURE))
            pendingRequest = null
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != REQ_RECORD_AUDIO) return
        val req = pendingRequest ?: return
        pendingRequest = null

        val granted = grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        runOnUiThread {
            if (granted) {
                Log.i(TAG, "Android RECORD_AUDIO granted → granting AUDIO_CAPTURE to WebView")
                req.grant(arrayOf(PermissionRequest.RESOURCE_AUDIO_CAPTURE))
            } else {
                Log.w(TAG, "Android RECORD_AUDIO denied → denying WebView request")
                req.deny()
            }
        }
    }

    override fun onDestroy() {
        pendingRequest = null
        super.onDestroy()
    }

    private fun isOriginAllowed(origin: String): Boolean {
        if (origin.isBlank()) return false
        return ALLOWED_ORIGINS.any { allowed ->
            origin == allowed || origin.startsWith("$allowed/") || origin.startsWith("$allowed:")
        }
    }
}
